cohere/client.py

```diff
@@ -503,6 +503,7 @@ def __request(self, endpoint, json=None, method='POST') -> Any:
                     backoff_factor=0.5,
                     allowed_methods=['POST', 'GET'],
                     status_forcelist=[429, 500, 502, 503, 504],
+                    raise_on_status=False,
```

Comment:
I think what we should go is still raise on status but then catch the dofferent error types. The issue with the current implementation is that it assumes that you got a response back, but if you have a connection error it will crap out

Reply:
In the 3.9 branch this change seems fine, but it would be good if @lfayoux could go over the setup in the 4.0 version as well to see what the effects are.

Reply:
If we raise on status and catch an error then we would lose the response message that we get back. I can a try to the request to catch connection errors though

Reply:
i think we should do it like this:

```python
with requests.Session() as session:
    retries = Retry(
        total=3,
        backoff_factor=0.5,
        allowed_methods=['POST', 'GET'],
        status_forcelist=[429, 500, 502, 503, 504],
        raise_on_status=False,
    )
    session.mount('https://', HTTPAdapter(max_retries=retries))
    session.mount('http://', HTTPAdapter(max_retries=retries))

try:
    response = session.request(method, ...)
    response.raise_for_status()
except requests.exceptions.HTTPError as http_error:
    # check response.status_code and handle appropriately
except requests.exceptions.ConnectionError as connection_error:
    # Handle connection error
except requests.exceptions.Timeout as timeout_error:
    # Handle timeout error
except requests.exceptions.RequestException as request_error:
    # Handle any other error
```